File: src/test/java/com/flexpoker/table/command/aggregate/pot/BasicPotHandlerTest.java
package com.flexpoker.table.command.aggregate.pot;

import com.flexpoker.exception.FlexPokerException;
import org.junit.jupiter.api.Test;

import java.util.NoSuchElementException;
import java.util.UUID;

import static com.flexpoker.table.command.aggregate.pot.PotTestUtils.createBasicPotHandler;
import static com.flexpoker.table.command.aggregate.pot.PotTestUtils.createSetOfPlayers;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class BasicPotHandlerTest {

    @Test
    void testTwoPlayersRequiredToShowCardsAndChipsWon() {
        var player1 = UUID.randomUUID();
        var player2 = UUID.randomUUID();
        var potId1 = UUID.randomUUID();
        var potId2 = UUID.randomUUID();
        var playersInvolved = createSetOfPlayers(player1, player2);

        var potHandler = createBasicPotHandler(player1, player2);
        potHandler.addNewPot(potId1, playersInvolved);
        potHandler.addNewPot(potId2, playersInvolved);
        potHandler.addToPot(potId1, 20);
        potHandler.addToPot(potId2, 10);

        var playersRequiredToShowCards = potHandler.fetchPlayersRequiredToShowCards(playersInvolved);
        var fetchChipsWon = potHandler.fetchChipsWon(playersInvolved);

        assertTrue(playersRequiredToShowCards.contains(player1));
        assertFalse(playersRequiredToShowCards.contains(player2));
        assertEquals(30, fetchChipsWon.get(player1).intValue());
    }

    @Test
    void testAddNewPotWithPlayersThatDontExist() {
        var player1 = UUID.randomUUID();
        var player2 = UUID.randomUUID();
        var player3 = UUID.randomUUID();

        var potHandler = createBasicPotHandler(player1, player2);
        
        var potId = UUID.randomUUID();

        assertThrows(IllegalArgumentException.class,
                () -> potHandler.addNewPot(potId, createSetOfPlayers(player3)));
    }

    @Test
    void testClosePot() {
        var player1 = UUID.randomUUID();
        var player2 = UUID.randomUUID();

        var potHandler = createBasicPotHandler(player1, player2);
        
        var potId = UUID.randomUUID();
        potHandler.addNewPot(potId, createSetOfPlayers(player1, player2));
        potHandler.closePot(potId);
    }

    @Test
    void testClosePotThatDoesntExist() {
        PotHandler potHandler = createBasicPotHandler(UUID.randomUUID(), UUID.randomUUID());
        assertThrows(NoSuchElementException.class,
                () -> potHandler.closePot(UUID.randomUUID()));
    }

    @Test
    void testAddChipsToClosedPot() {
        var player1 = UUID.randomUUID();
        var player2 = UUID.randomUUID();

        var potHandler = createBasicPotHandler(player1, player2);
        
        var potId = UUID.randomUUID();
        potHandler.addNewPot(potId, createSetOfPlayers(player1, player2));
        potHandler.closePot(potId);

        assertThrows(FlexPokerException.class,
                () -> potHandler.addToPot(potId, 10));
    }

}

File: src/test/kotlin/com/flexpoker/table/command/aggregate/pot/BasicPotHandlerTest.kt
package com.flexpoker.table.command.aggregate.pot

import com.flexpoker.table.command.aggregate.addNewPot
import com.flexpoker.table.command.aggregate.addToPot
import com.flexpoker.table.command.aggregate.closePot
import com.flexpoker.table.command.aggregate.fetchChipsWon
import com.flexpoker.table.command.aggregate.fetchPlayersRequiredToShowCards
import com.flexpoker.table.command.aggregate.removePlayerFromAllPots
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.pcollections.HashTreePSet
import java.util.UUID

class BasicPotHandlerTest {

    @Test
    fun testTwoPlayersRequiredToShowCardsAndChipsWon() {
        val player1 = UUID.randomUUID()
        val player2 = UUID.randomUUID()
        val potId1 = UUID.randomUUID()
        val potId2 = UUID.randomUUID()
        val playersInvolved = setOf(player1, player2)
        val hands = createPotHands(player1, player2)
        var pots = addNewPot(HashTreePSet.empty(), hands, potId1, playersInvolved)
        pots = addNewPot(pots, hands, potId2, playersInvolved)
        pots = addToPot(pots, potId1, 20)
        pots = addToPot(pots, potId2, 10)
        val playersRequiredToShowCards = fetchPlayersRequiredToShowCards(pots, playersInvolved)
        val fetchChipsWon = fetchChipsWon(pots, playersInvolved)
        assertTrue(playersRequiredToShowCards.contains(player1))
        assertFalse(playersRequiredToShowCards.contains(player2))
        assertEquals(30, fetchChipsWon[player1]!!.toInt())
    }

    @Test
    fun testAddNewPotWithPlayersThatDontExist() {
        val player1 = UUID.randomUUID()
        val player2 = UUID.randomUUID()
        val player3 = UUID.randomUUID()
        val hands = createPotHands(player1, player2)
        val potId = UUID.randomUUID()
        assertThrows(IllegalArgumentException::class.java) { addNewPot(HashTreePSet.empty(), hands, potId, setOf(player3)) }
    }

    @Test
    fun testClosePot() {
        val player1 = UUID.randomUUID()
        val player2 = UUID.randomUUID()
        val hands = createPotHands(player1, player2)
        val potId = UUID.randomUUID()
        var pots = addNewPot(HashTreePSet.empty(), hands, potId, setOf(player1, player2))
        pots = closePot(pots, potId)
        assertFalse(pots.first { it.id == potId }.isOpen)
    }

    @Test
    fun testClosePotThatDoesntExist() {
        assertThrows(IllegalArgumentException::class.java) { closePot(HashTreePSet.empty(), UUID.randomUUID()) }
    }

    @Test
    fun testAddChipsToClosedPot() {
        val player1 = UUID.randomUUID()
        val player2 = UUID.randomUUID()
        val hands = createPotHands(player1, player2)
        val potId = UUID.randomUUID()
        var pots = addNewPot(HashTreePSet.empty(), hands, potId, setOf(player1, player2))
        pots = closePot(pots, potId)
        assertThrows(IllegalArgumentException::class.java) { addToPot(pots, potId, 10) }
    }

    @Test
    fun testPotIsOpenOnCreate() {
        val player1 = UUID.randomUUID()
        val player2 = UUID.randomUUID()
        val hands = createPotHands(player1, player2)
        val potId = UUID.randomUUID()
        var pots = addNewPot(HashTreePSet.empty(), hands, potId, setOf(player1, player2))
        assertEquals(1, pots.size)
        assertTrue(pots.first().isOpen)
    }

    @Test
    fun testPotCannotHaveAPlayerRemovedAfterClosing() {
        val player1 = UUID.randomUUID()
        val player2 = UUID.randomUUID()
        val hands = createPotHands(player1, player2)
        val potId = UUID.randomUUID()
        var pots = addNewPot(HashTreePSet.empty(), hands, potId, setOf(player1, player2))
        pots = closePot(pots, potId)
        assertThrows(IllegalArgumentException::class.java) { removePlayerFromAllPots(pots, player1) }
    }

    @Test
    fun testPotPlayerRemovedHitsRequire() {
        val player1 = UUID.randomUUID()
        val hands = createPotHand(player1)
        val potId = UUID.randomUUID()
        var pots = addNewPot(HashTreePSet.empty(), listOf(hands), potId, setOf(player1))
        pots = closePot(pots, potId)
        assertThrows(IllegalArgumentException::class.java) { removePlayerFromAllPots(pots, player1) }
    }

    @Test
    fun testPotCanHaveAPlayerRemoved() {
        val player1 = UUID.randomUUID()
        val player2 = UUID.randomUUID()
        val hands = createPotHands(player1, player2)
        val potId = UUID.randomUUID()
        var pots = addNewPot(HashTreePSet.empty(), hands, potId, setOf(player1, player2))
        pots = removePlayerFromAllPots(pots, player1)
        val p1Pot = pots.first().handEvaluations.firstOrNull { it.playerId == player1 }
        val p2Pot = pots.first().handEvaluations.firstOrNull { it.playerId == player2 }
        assertNull(p1Pot)
        assertNotNull(p2Pot)
    }

}